tools/wake-format/README.md

```diff
@@ -0,0 +1,481 @@
+# wake-format
+
+`wake-format` is a tool for automatically applying the wake style guide to wake source files. It is intentionally opinionated to create a singular correct formatting standard.
+
+# Usage
+
+The standard usage of `wake-format` is to format a set of wake files in place.
+
+`wake-format -i source.wake source2.wake source3.wake`
+
+`wake-format` supports other flows for various use cases. The `-n`/`--dry-run` flag is ideal for CI/CD to check if the source needs to be formatted. It exits with a non-zero status code when a file would be changed by `wake-format`
+
+Running `wake-format` without `-i` will emit the file to `stdout` instead of overwritting it.
+
+# Documentation
+
+The information below should be helpful to contribute to `wake-format`
+
+## Building
+
+`wake-format` is built in the same manner as `wake`. Running `make` or `./bin/wake build default` from within the wake dir will produce the `wake-format` binary.
+
+The locally build `wake-format` can be found at `./bin/wake-format.native-cpp14-release`
+
+## wake-format Fluent API/DSL
+
+`wake-format` relies heavily on a custom fluent API/DSL to generate the formatting rules. Usage of the API is documented below. See [here](TODO) for design decisions, invariants, and engineering tradeoffs considered.
+
+### Usage
+
+All formatting rules start by creating a basic formatting object using `fmt()`. From there functions are added to describe the expectations of the current [CST node](TODO) and to return the appropiate text.
+
+```c++
+auto basic_fmt = fmt()
+    .space()
+    .newline();
+```
+
+creates a formatter that returns a space and a newline.
+
+Once a formatter has been defined, the `format()` function is used to run the formatter and return the formated source. The source is returned as a [wcl::doc](TODO) for efficient concatenation and geometry lookup. `format()` takes a `ctx_t` and a `CSTElement`. The `ctx_t` (context) tracks the context (such as nest level) that all subnodes should be formatted in, and the `CSTElement` is the current node being formatted.
+
+```c++
+ctx_t ctx;
+CSTElement node = ""; // An empty node
+
+wcl::doc doc = fmt()
+  .space()
+  .newline()
+  .format(ctx, node);
+
+std::string s = doc.as_string(); // " \n"
+```
+
+`format()` expects to consume the entire `CSTElement` node and will `assert()` if that is not the case. Asserting may seem heavily handed, however this is a very powerful method for identifying and fixing nodes with unexpected contents. 
+
+### API Functions & Examples
+
+**CSTElement Manipulation**
+
+Manipulation of the current state of the formatter relating to CSTElement nodes.
+
+**`token(uint8_t id)`**
```

Comment:
Is it important that the user knows that it's a `uint8_t`, wondering if a typedef (or other opaque) would make it more self describing

Reply:
Jake and I talked about this previously. I'm actually very pro adding a typedef here. Would make it easier to bump the size if we ever have more that 256 IDs as well. 

It's one of those things where I'd want to change it everywhere if we added it though. I'll look into at least making the change for `wake-format` but I think `wake` would greatly benefit from it